pkg/cluster/operation/upgrade.go

```diff
@@ -142,3 +179,86 @@ func Addr(ins spec.Instance) string {
 	}
 	return ins.GetHost() + ":" + strconv.Itoa(ins.GetPort())
 }
+
+var (
+	leaderScheduleLimitOffset = 32
+	regionScheduleLimitOffset = 512
+	//storeLimitOffset             = 512
+	leaderScheduleLimitThreshold = 64
+	regionScheduleLimitThreshold = 1024
+	//storeLimitThreshold          = 1024
+)
+
+// increaseScheduleLimit increases the schedule limit of leader and region for faster
+// rebalancing during the rolling restart / upgrade process
+func increaseScheduleLimit(ctx context.Context, pc *api.PDClient) (
+	currLeaderScheduleLimit int,
+	currRegionScheduleLimit int,
+	err error) {
+	// insert checkpoint
+	point := checkpoint.Acquire(ctx, map[string]interface{}{
+		"operation": "upgrade",
+		"func":      "increaseScheduleLimit",
+	})
+	defer func() {
+		point.Release(err,
+			zap.String("operation", "upgrade"),
+			zap.String("func", "increaseScheduleLimit"),
+			zap.Int("currLeaderScheduleLimit", currLeaderScheduleLimit),
+			zap.Int("currRegionScheduleLimit", currRegionScheduleLimit),
+		)
+	}()
+
+	if data := point.Hit(); data != nil {
+		currLeaderScheduleLimit = data["currLeaderScheduleLimit"].(int)
+		currRegionScheduleLimit = data["currRegionScheduleLimit"].(int)
+		return
+	}
+
+	// query current values
+	cfg, err := pc.GetConfig()
+	if err != nil {
+		return
+	}
+	val, ok := cfg["schedule.leader-schedule-limit"].(float64)
+	if !ok {
+		return currLeaderScheduleLimit, currRegionScheduleLimit, perrs.New("cannot get current leader-schedule-limit")
+	}
+	currLeaderScheduleLimit = int(val)
+	val, ok = cfg["schedule.region-schedule-limit"].(float64)
+	if !ok {
+		return currLeaderScheduleLimit, currRegionScheduleLimit, perrs.New("cannot get current region-schedule-limit")
+	}
+	currRegionScheduleLimit = int(val)
+
+	// increase values
+	if currLeaderScheduleLimit < leaderScheduleLimitThreshold {
+		newLimit := currLeaderScheduleLimit + leaderScheduleLimitOffset
+		if newLimit > leaderScheduleLimitThreshold {
+			newLimit = leaderScheduleLimitThreshold
+		}
+		if err := pc.SetReplicationConfig("leader-schedule-limit", newLimit); err != nil {
+			return currLeaderScheduleLimit, currRegionScheduleLimit, err
+		}
+	}
+	if currRegionScheduleLimit < regionScheduleLimitThreshold {
+		newLimit := currRegionScheduleLimit + regionScheduleLimitOffset
+		if newLimit > regionScheduleLimitThreshold {
+			newLimit = regionScheduleLimitThreshold
+		}
+		if err := pc.SetReplicationConfig("region-schedule-limit", newLimit); err != nil {
+			return currLeaderScheduleLimit, currRegionScheduleLimit, err
```

Comment:
do we need rollback `leader-schedule-limit` here?

Maybe changes L197 `err error` -> `retErr err` a defer on L 242:

```go
defer func() {
  if retErr != nil {
    pc.SetReplicationConfig("leader-schedule-limit", currLeaderScheduleLimit)
  }
}()
```

Reply:
If setting it to a new value fails, I don't think we need to try to set it to the old value again.

Reply:
I mean we should revert `leader-schedule-limit` when setting `region-schedule-limit` faild